File: go/p4/p4_v1.go
package main

import (
	"fmt"
	"strconv"
	"time"
)

func problem4() int {
	var isPalindrome = func(n int) bool {
		nAsString := strconv.Itoa(n)
		for i, j := 0, len(strconv.Itoa(n))-1; i < j; i, j = i+1, j-1 {
			if nAsString[i] != nAsString[j] {
				return false
			}
		}
		return true
	}

	const upperBound int = 999
	const lowerBound int = 99

	palindrome := 0

	multiplicand := upperBound

	for multiplicand > lowerBound && multiplicand*upperBound > palindrome {
		for multiplier := upperBound; multiplier > lowerBound; multiplier-- {
			product := multiplicand * multiplier
			if product > palindrome && isPalindrome(product) {
				palindrome = product
			}
		}
		multiplicand--
	}

	return palindrome
}

func main() {
	start := time.Now()
	fmt.Println("Result:", problem4())
	elapsed := time.Since(start)
	fmt.Printf("Execution time: %.6f seconds\n", elapsed.Seconds())
}

File: python/p4/p4_v1.py
def problem4():
    upper_bound = 999
    lower_bound = 99

    palindrome = 0

    number = upper_bound
    while number > lower_bound and number * upper_bound > palindrome:
        prime = upper_bound
        while prime > lower_bound:
            if str(number * prime) == str(number * prime)[::-1] and number * prime > palindrome:
                palindrome = number * prime
            prime -= 1
        number -= 1
    return palindrome

if __name__ == "__main__":
    import time
    
    start_time = time.time()
    result = problem4()
    end_time = time.time()

    print(f"Result: {result}")
    print(f"Execution time: {end_time - start_time:.6f} seconds")